src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+
+        public void resume ( )
+        {
+            if (this.fiber_suspended == fiber_suspended.RequestSuspended)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Requests the fiber to terminate when all remaining records have been
+            passed to the user. Called when a `Stopped` message from the node
+            has arrived.
+
+        ***********************************************************************/
+
+        public void stop ( )
+        {
+            this.stopped = true;
+            if (this.fiber_suspended == fiber_suspended.WaitingForRecords)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Waits for a batch of records to arrive and passes it to the user,
+            then sends the `Continue` message to the node, in a loop. Handles
+            suspending the request through the controller, for resuming call
+            `resume`. Calling `stop` makes this routine terminate after all
+            remaining records have been passed to the user.
+
+        ***********************************************************************/
+
+        private void fiberMethod ( )
+        {
+            while (this.waitForRecords())
+            {
+                for (uint yield_count = 1; ; yield_count++)
+                {
+                    if (this.outer.context.shared_working.suspendable_control.suspended)
+                    {
+                        yield_count = 0;
+                        this.fiber_suspended = fiber_suspended.RequestSuspended;
+                        this.fiber.suspend(token, this);
+                        this.fiber_suspended = fiber_suspended.No;
+                    }
+
+                    this.passRecordToUser(
+                        this.outer.conn.message_parser.getArray
+                        !(Const!(void))(this.remaining_batch)
+                    );
+
+                    if (!this.remaining_batch.length)
+                        break;
+
+                    if (yield_count >= 10) // yield every 10th record
+                    {
+                        yield_count = 0;
+                        this.outer.request_event_dispatcher.yield(this.fiber);
+                    }
+                }
+
+                this.remaining_batch = null;
+                (*this.batch_buffer).length = 0;
+
+                if (this.stopped)
+                    break;
+
+                this.outer.request_event_dispatcher.send(
+                    this.fiber,
+                    (conn.Payload payload)
+                    {
+                        payload.addConstant(MessageType_v3.Continue);
+                    }
+                );
+            }
+
+            this.outer.request_event_dispatcher.signal(this.outer.conn,
```

Comment:
Did you try using `abort` here instead? It seems like that should do the same job and wouldn't require this extra signal code.

Reply:
Oh I see, but then you'd need the controller fiber. This is why I ended up jumping through hoops to make all scope classes accessible to each other. Interesting. I guess both approaches have pros and cons.

Actually, a nice middle way might be to add the *fibers* as members of `ConsumeHandler`, thus accessible to all of the nested scope classes.